### PR TITLE
Add Debian packaging

### PR DESCRIPTION
**Pull request by mcnewton**
Hi,

I wanted to put this (great looking!) software on a couple of servers for testing. However, we need it in Debian packages for that and there was no debian/ directory, so I've created basic Debian packaging for it.

It's not entirely lintian clean; the main issue being no init script so Jessie only (and a few things such as sources missing for minified js files, which probably should be pulled in from other Debian packages per normal Debian policy), but it's fine for use on a local system.

I added in the standard configure script, which was missing, so that there is no build dependency on the autoconf/automake tools. Systems shouldn't generally be required to have these installed to be able to build.

The packaging puts the systemd service file on as well as a logrotate script. As the web files are user-editable it puts them in /var/lib/netdata owned by a new netdata user which it creates/removes.

I think the only possible controversial thing it does is to chmod 4755 the apps plugin. Your installer does this, but I think it's generally frowned upon in Debian circles because of the security risk (might be wrong, though). Maybe the daemon could keep enough root privileges to be able to run plugins as root when required, rather than making them setuid and accessible by any user of the system? The normal Debian solution would probably be to install the file non-suid (so lose the functionality) but mention it in the Debian README.

Aside from that, it seems to work fine here. Hope this is useful!

Cheers,
Matthew

### CODE REVIEW

**Comment by cbricart** [comment]
this is the live Repository you're committing this to.. **that's why** you all need the "build dependency on autotools"(!!)
**released** tarballs **will not** [and] have [not] this build dependency..

**Comment by mcnewton** [comment]
OK, thanks. I'm aware this is the master repository.

In projects I usually work on I'm used to the build tools being available in both the git repository and the released tarballs. This makes it easy for anyone to build wherever they got the source from. It also means packaging (for whatever system) will be easier because you don't need to generate configure before building the package (and therefore have the extra dependency).

But in that case the solution if you don't like having the configure script is to just pull that one commit out. The package build process could probably then run autogen.sh at the start if configure doesn't already exist.

Released version 1.0.0 at https://github.com/firehol/netdata/releases doesn't contain a configure script so you might want to fix that in later releases if that's the intended position?

Cheers,
Matthew

**Comment by cbricart** [comment]
to quote the releases page here:
> netdata 1.0.0 - download release tarfiles from http://firehol.org/download/netdata/releases/v1.0.0
 ;)

**Comment by cbricart** [comment]
to quote the releases page here on GH:
> download release tarfiles from http://firehol.org/download/netdata/releases/v1.0.0

;)

**Comment by mcnewton** [comment]
Ah! That releases page is confusing (to me, at least)...

It says in large letters "v1.0.0" then "Source code" then has a zip file and tarball... which don't contain the configure script. So another reason why having the configure script available to github makes it easier for people. :-)

Anyway... I've just updated this the PR to remove the configure script and to build configure if it doesn't exist as part of the packaging process. Maybe that is better?

Cheers,
Matthew

**Comment by ktsaou** [comment]
You are so nice!

One last thing: to have these files in the release tar, we need to add some kind of Makefile.am
But I am not an expert for this.

@alonbl , @philwhineray can you review this please?

**Comment by mcnewton** [comment]
Just pushed a new update to this PR. It now builds clean in a jessie sbuild environment, so it should be good on any jessie system. Required more build dependencies because it has to run autoconf first :(. But it's worked around now.

I've also made it source format 3.0 (native). Native won't be suitable for the main Debian repository, but I generally find it's better in a project source because it means packages can built easier without the worry of having to have a source tarball around to create diffs from (there's no point creating a diff - this _is_ the original source). If this gets put into Debian proper then the maintainer there will set that back to plain 3.0 and worry about the source tarball :)

Cheers,
Matthew

**Comment by ktsaou** [comment]
I am planning for a new release tomorrow.
Shall we include this?

**Comment by philwhineray** [comment]
@mcnewton we don't include the configure because it is generated and risks getting out of step. The netdata project does not include PDFs (yet?) but other projects do and they need building at some point - the process is the same... it gets packaged into a release tar-file so a majority of people do not need the extra dependencies.

The confusion stems because I don't know how (or if it is possible) to suppress the links that github generates to itself whenever a git tag is made. The best I know to do is put some text above them (as we do) with a link to the preferred location.

@ktsaou I have been asked by debian developers in the past to not include a `debian/` directory in the base of the package because this then interferes with their making an "official" package. @lhw has announced he Intends To Package for debian in #42 and if that is still the case, as a debian developer it is probably best to defer to his preferences.

**Comment by mcnewton** [comment]
Hi,

It would just be good to add an init script for wheezy systems if possible. But aside from that it should be a good start. I'm not able to look at that until at least tomorrow evening, though, sorry.

My personal view on software is generally to make things as easy as possible for the end users who might want to get it installed. Hence providing a configure script in the source and a debian dir in the correct location. This matters especially when projects are new or in fast development and people are more likely to want to get it from the original source. And when it gets into Debian proper it will only get updated once every couple of years (e.g. we have no end of problems with FreeRADIUS because of questions about obsolete versions Debian provide on the mailing lists). Developers are smart enough to script things (such as rerun autogen on every commit) whereas users find unusual things harder and it counts against the project.

However, _please_ don't think I'm trying to make you change your project if you want to do it the current way! Just explaining my thoughts. :-)

Another project I submitted Debian packaging to just put it in a contrib directory or similar. That might be a way to keep people happy without being too hard? Just need to move into place before building.

Cheers!
Matthew

**Comment by philwhineray** [comment]
I think putting something in a `contrib` directory would be a fine solution, especially if this can then be used by OBS or similar to produce installable binaries for the project which will make things easier for the end user.

I totally understand the desire to be able to get packages on a separate schedule to the one the distros follow I just don't want to make their lives harder than necessary either.

File: charts.d/ap.chart.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (alonbl) [inline charts.d/ap.chart.sh]: @ktsaou: can't we remove bashism?

File: debian/rules [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+override_dh_auto_configure:
+	# Build the configure script if it doesn't already exist
+	-[ ! -x $(CURDIR)/configure ] && $(CURDIR)/autogen.sh
+	dh_auto_configure -- --with-math --with-webdir=/var/lib/netdata
+
+override_dh_install:
+	dh_install
+	# Documentation and JavaScript should not be executable
+	chmod 644 $(CURDIR)/debian/netdata/usr/lib/*/netdata/*/*.js
+	chmod 644 $(CURDIR)/debian/netdata/usr/lib/*/netdata/*/*.md
```
Comment (alonbl) [inline debian/rules]: if we install these as executables, we need to fix the make install.

Comment (alonbl) [inline debian/rules]: can you please try to rebase on top of #141 and see if it helps?

File: debian/rules [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	# This config file is also a script
+	chmod 755 $(CURDIR)/debian/netdata/etc/netdata/charts.d.conf
```
Comment (alonbl) [inline debian/rules]: should this be run using '.' (source)? if not, we should modify Makefile.am to dist_config_SCRIPTS and remove this.

File: debian/rules [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	# Remove unneeded files
+	rm $(CURDIR)/debian/netdata/var/cache/netdata/.keep
+	rm $(CURDIR)/debian/netdata/var/log/netdata/.keep
```
Comment (alonbl) [inline debian/rules]: you can just do same as we do in spec, so you won't change this in future:
```
find "$(CURDIR)/debian" -name .keep -exec rm {} \;
```

File: debian/rules [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+override_dh_installinit:
+	# No need to duplicate the service file again, so use a copy
+	cp system/netdata-systemd debian/netdata.service
```
Comment (alonbl) [inline debian/rules]: Note: I modified this to netdata.systemd in #140.

File: charts.d/ap.chart.sh (hunk 1 above)
Comment (ktsaou) [inline charts.d/ap.chart.sh]: > can't we remove bashism?

mysql and opensips depend heavily on it.

I am afraid it won't be easy (mainly due to arrays). In fact I prefer to re-write most BASH plugins to another language (node, python, whatever).

In all cases though, shell plugins should be an available option and the only way to achieve acceptable speeds with shell, is BASH. (I know BASH is the slowest of all shells, however its arrays make it especially userful and speedy for repeating tasks, compared to any other shell).

**Comment by mcnewton** [comment]
@alonbl thanks for all the great suggestions! I'll be back home later today and will try to update this as soon as I can then.

File: debian/rules (hunk 2 above)
Comment (mcnewton) [inline debian/rules]: Much better thanks!

Please see #166 for the few remaining permission errors.

File: debian/rules (hunk 3 above)
Comment (mcnewton) [inline debian/rules]: I have updated this in #166 so this line can be removed, but if it is to be sourced instead then the shebang should really be removed.

File: debian/rules (hunk 5 above)
Comment (mcnewton) [inline debian/rules]: (netdata.service) - thanks, updated.

File: debian/rules (hunk 3 above)
Comment (alonbl) [inline debian/rules]: I updated the #141 please see if it is now ok, removed the shebang from config as indeed it seems to be sourced and not executed.

**Comment by alonbl** [comment]
@mcnewton: why have you closed this?

**Comment by mcnewton** [comment]
Arrr, git push typo accidentally closed this. Sorry.

Is it possible to re-open it or do I have to do another one?

**Comment by mcnewton** [comment]
I was about to push the latest updates.

**Comment by alonbl** [comment]
you should use the same branch name and force push, but because you deleted the branch then I cannot reopen this request.

**Comment by mcnewton** [comment]
A warning not o "git push github **:**debianpkg" by mistake :-(. Or to not try and do stuff when you've got a cold and can't breathe properly. Not sure which...

I've pushed the original debianpkg branch back with exactly the same HEAD as before - is it possible to reopen now? Otherwise I guess I'll just have to do another PR :(

**Comment by mcnewton** [comment]
Ah, cool - updated this page and reopen appeared :-)

**Comment by alonbl** [comment]
yes, another one...

**Comment by alonbl** [comment]
nice!

**Comment by mcnewton** [comment]
Right, let's try that again....

* moved to contrib/ @philwhineray
* removed chmods from debian/rules
* changed rm .../.keep to find -exec rm
* updated systemd service file name from #140 

Lintian errors now are:
* fonts in other debian package (warning, can ignore for the time being)
* suid binary (not sure the best way to handle this, but still only a warning for now)
* **missing init script** (error, policy violation)
* netdata binary doesn't have manpage (warning, can ignore for now)

I'll see if I can knock up a quick init script...

File: contrib/debian/netdata.logrotate [hunk 6]
```diff
@@ -0,0 +1,13 @@
+/var/log/netdata/*.log {
```
Comment (alonbl) [inline contrib/debian/netdata.logrotate]: this has nothing to do with debian, please submit as separate patch into system/

Comment (alonbl) [inline contrib/debian/netdata.logrotate]: also make this a template so /var/log will be the @whatever@

File: netdata-9999.ebuild [hunk 7]
```diff
@@ -8,38 +8,33 @@ inherit linux-info systemd user
...
 else
 	SRC_URI="http://firehol.org/download/${PN}/releases/v${PV}/${P}.tar.xz"
+	KEYWORDS="~amd64 ~x86"
 fi
 
 
-DESCRIPTION="Linux real time system monitoring, over the web!"
-HOMEPAGE="https://github.com/firehol/netdata"
+DESCRIPTION="Linux real time system monitoring, done right!"
+HOMEPAGE="https://github.com/firehol/netdata http://netdata.firehol.org/"
```
Comment (alonbl) [inline netdata-9999.ebuild]: I think the github link should be sufficient.

File: netdata-9999.ebuild [hunk 8]
```diff
@@ -59,34 +56,59 @@ pkg_setup() {
 	enewuser ${PN} -1 -1 / ${PN}
```
Comment (alonbl) [inline netdata-9999.ebuild]: if you defined NETDATA_USER, please use.

File: netdata-9999.ebuild [hunk 9]
```diff
@@ -59,34 +56,59 @@ pkg_setup() {
...
+		$(use_with compression zlib)
 }
 
 src_install() {
 	default
 
-	fowners ${PN} /var/log/netdata
+	fowners ${NETDATA_USER} /var/log/netdata
+
+	chown -Rc ${NETDATA_USER} "${ED}"/usr/share/${PN} || die
```
Comment (alonbl) [inline netdata-9999.ebuild]: this should not be required

File: system/netdata-openrc.in [hunk 10]
```diff
@@ -0,0 +1,83 @@
+#!/sbin/runscript
+
+# The user netdata is configured to run as.
+# If you edit its configuration file to set a different
+# user, set it here too, to have its files switch ownership
+: ${NETDATA_OWNER:=netdata:netdata}
```
Comment (alonbl) [inline system/netdata-openrc.in]: we should expose the user out of the autoconf.

Comment (alonbl) [inline system/netdata-openrc.in]: we deal with this file at #166, I guess better to leave it as-is to avoid conflicts for now.

File: netdata-9999.ebuild (hunk 9 above)
Comment (mcnewton) [inline netdata-9999.ebuild]: Hmm, that's not part of this patch. looks like PR has sucked in another commit. I'll sort that out!

File: system/netdata-openrc.in (hunk 10 above)
Comment (mcnewton) [inline system/netdata-openrc.in]: Ditto to previous.

File: contrib/debian/netdata.logrotate (hunk 6 above)
Comment (mcnewton) [inline contrib/debian/netdata.logrotate]: This is now templated in system/ and a separate commit.

File: system/Makefile.am [hunk 11]
```diff
@@ -5,6 +5,7 @@ MAINTAINERCLEANFILES= $(srcdir)/Makefile.in
 CLEANFILES = \
 	netdata-openrc \
 	netdata.service \
+	netdata.logrotate \
```
Comment (alonbl) [inline system/Makefile.am]: please sort() files...

**Comment by mcnewton** [comment]
Hopefully looking a bit better now. Logrotate script is now in system/netdata.logrotate. Rebased against master (again) which has removed the erroneous commit included before (always love rebase... does nearly what you want it to, but always with a surprise!).

File: system/Makefile.am [hunk 12]
```diff
@@ -14,9 +15,11 @@ SUFFIXES = .in
 nodist_noinst_DATA = \
 	netdata-openrc \
 	netdata.service \
+	netdata.logrotate \
```
Comment (alonbl) [inline system/Makefile.am]: same, sorry.

File: system/Makefile.am (hunk 11 above)
Comment (mcnewton) [inline system/Makefile.am]: that's just picky ;-P

done.

File: system/netdata.logrotate.in [hunk 13]
```diff
@@ -0,0 +1,13 @@
+@localstatedir_POST@/log/netdata/*.log {
+	daily
+	missingok
+	rotate 14
+	compress
+	delaycompress
+	notifempty
+	sharedscripts
+	postrotate
+		systemctl restart netdata.service
```
Comment (alonbl) [inline system/netdata.logrotate.in]: oh! sorry, I did not notice that before.
do we must use systemd specific commands? and restart?
worse case logrotate should use the mode of copytruncate.

File: system/Makefile.am (hunk 11 above)
Comment (alonbl) [inline system/Makefile.am]: it helps to avoid conflicts...

File: system/Makefile.am [hunk 14]
```diff
@@ -3,6 +3,7 @@
 #
 MAINTAINERCLEANFILES= $(srcdir)/Makefile.in
 CLEANFILES = \
+	netdata.logrotate \
```
Comment (alonbl) [inline system/Makefile.am]: I am sooooooooo sorry... but this is not sorted... should be before the netdata.service, if you use vi, just select the lines and :sort

**Comment by mcnewton** [comment]
Getting closer... I've just added an init script locally, will push when I've tested it (need to find a wheezy system).

File: system/Makefile.am (hunk 14 above)
Comment (mcnewton) [inline system/Makefile.am]: It is according to 'ls' here...

$ ls system/ | grep in
Makefile.in
netdata.logrotate.in
netdata-openrc.in
netdata.service.in
$

but I guess that's using the locale. <sigh> I'll update it again so it's ascii-sorted instead.....

File: system/netdata.logrotate.in (hunk 13 above)
Comment (mcnewton) [inline system/netdata.logrotate.in]: Yeah, looking at that as part of adding an init script, as it's no good for wheezy.
copytruncate is hideous, /etc/init.d/netdata restart should work on most things though (even Debian jessie, where it will be intercepted and translated to a systemd call)

Comment (shartge) [inline system/netdata.logrotate.in]: You should never use `systemctl` or `/etc/init.d` directly but instead use the `service` wrapper:
`service netdata restart`

Comment (mcnewton) [inline system/netdata.logrotate.in]: Good point! Thanks! :)

**Comment by mcnewton** [comment]
Pushed latest version...

* Now has init script, and therefore passes lintian checks :)
* Will build and run on Debian wheezy. Needs a couple of systemd-related tweaks - see debian/rules. This should probably just be mentioned on a wiki page somewhere if someone is still running wheezy and trying to build for that.
* /etc/default/netdata added so parameters can be passed. systemd unit file diverged enough from (and not compatible with) the redhat one so just put in the debian dir as would normally be the case anyway. Pure jessie/stretch way I believe is to just use a systemd override file, but /etc/defaults it is for the time being and backwards compatibility.
* logrotate script uses service (thanks @shartge!). tested and works fine on wheezy and jessie.

Go bug hunting... :)

**Comment by shartge** [comment]
May I propose a simpler systemd unit? There is no need for a separate ExecStop or even a PidFile when using systemd. Just have netdata stay in the foreground and systemd will do the rest.

I have been using the following unit for the last 3 days without problems:

```
[Unit]
Description=netdata real-time system monitoring
After=network.target nfs-kernel-server.service apache2.service mysql.service exim4.service bind9.service

[Service]
Type=simple
ExecStart=/usr/sbin/netdata -nd
TimeoutStopSec=30
Restart=5s

[Install]
WantedBy=multi-user.target
```

**Comment by mcnewton** [comment]
Agreed, as long as netdata is fine running in the foreground then using simple should be good.

I would be inclined to leave the EnvironmentFile / EXTRA_OPTS bits in though, personally. systemd is annoying enough for many people as it is without making some other long-standing configuration feature break :-)

**Comment by shartge** [comment]
Sure, but what EXTRA_OPTS does one need to provide that can not also be configured in the config.file?

**Comment by mcnewton** [comment]
Change the config file location... :)

**Comment by shartge** [comment]
Why?

**Comment by mcnewton** [comment]
Why not? I do this regularly on many systems. Push new config file out with cfengine in to, say, /opt/..., then update /etc/default/package to point to it. It makes the configuration management system cleaner and you don't get into trouble when you update packages.

**Comment by mcnewton** [comment]
But I agree netdata is good in that everything else is configurable in the config file. Some other software isn't so nice...!

**Comment by shartge** [comment]
You can do the same by providing a partial overlay unit for systemd in `/etc/systemd/system/netdata.service.d/configfile.conf`.
But if you want to provide the same options to the SysV init script and the systemd.unit, the environment _file is the most flexible option.

**Comment by mcnewton** [comment]
Yes, was just trying to keep things consistent with wheezy/jessie for now, and being kind to people who are very used to editing /etc/default/netdata :)

Especially as the package installs /etc/default/netdata anyway for the init script - so you then get people editing it and then wondering why it doesn't actually work. Though a comment in there to tell them how to override it with systemd is the other option I guess, but less clean IMO.

I've pushed it up with the slight modification to see what it looks like, but if people don't like it then I can get rid of it again.

**Comment by shartge** [comment]
Just be prepared for `EnvironmentFile` being removed from systemd in the future, because the developers really don't like that feature and think it is a mistake that they even implemented it.

**Comment by mcnewton** [comment]
:-(

I think it was a mistake that the developers wrote the rest of systemd as well, but let's not get in to that one here ;-) (though I'll admit it's often easier writing a service file than an init script, so I'll give them that one at least - until you can't do something and there's no workaround).

**Comment by shartge** [comment]
Right now I am pondering about the long list of services to start after. That is not really maintainable.
Question is: does netdata really need to start after all services to enable the data gatherers or is it able to dynamically enable them later?

If not, then the Type=simple should be changed to "Type=idle", which is the same as simple but delayed unitl all other jobs have been dispatched. Quite hackish though, IMHO.

**Comment by mcnewton** [comment]
Yes, I was wondering about the service list as well. It needs a "start at number 99" systemd equivalent. But from my limited experience I think it will just pick up new processes as they arrive, no? Maybe not for specific (mysql etc) plugins?

**Comment by shartge** [comment]
I'd say: remove them, but let's page @ktsaou to see what he has to say.

**Comment by mcnewton** [comment]
Hmm, another update that should be done is to move all the documentation into /usr/share/doc/netdata. The README.md files in the node/plugin/charts directories shouldn't be under /usr/lib for Debian packages. For tomorrow, now...

**Comment by ktsaou** [comment]
> remove them, but let's page @ktsaou to see what he has to say.

well, currently, if a plugin fails to gather data it will be disabled until the next scheduled re-run of the plugin. This scheduled re-run is controlled by the plugin itself:

- tc helper, will check for new interfaces every 2 minutes
- charts.d.plugin, restarts every 4 hours (configured by its `restart_timeout` option, in charts.d.conf)
- node.d.plugin, does not have this feature (it will run forever)
- apps.plugin, restarts every 4 hours (hardcoded in the binary)

Probably, the easiest could be for netdata to somehow to delay their start (a new command line option). Open a new issue and I'll what I can do it.

**Comment by shartge** [comment]
Delaying the start of a plugin won't really help, since you can't know how long the start of the rest of the system takes. For example I have one server which takes 10 minutes to be fully up after the network has been started.

**Comment by ktsaou** [comment]
> Delaying the start of a plugin won't really help, since you can't know how long the start of the rest of the system takes. For example I have one server which takes 10 minutes to be fully up after the network has been started.

This could be configurable - I was thinking of command line option to netdata.
Any ideas?

**Comment by mcnewton** [comment]
Thinking about how say other monitoring software, such as munin, manages this. There is a cron job weekly that detects which plugins should be enabled and enables them. ISTR this is done by running each plugin with a "--autodetect" type option.

Then the data collector just uses the plugins that are enabled. At boot time they all start (actually from cron every n minutes). If any fail, the graphs don't get updated. But they are always tried until the next weekly check would disable them.

It separates the plugin detection and plugin data collection. I think this is probably where the issue arises in trying to do both at the same time.

Agreed having a delay at the start just opens up to race conditions and would be best avoided. Could, however, say that if at startup no plugins are configured then autoconfigure the plugins. First time netdata is started is unlikely to be boot time so this would mean starting is seamless as it is now. Plugin detection could be from cron once a day (or even triggered from the netdata daemon itself rather than using cron, and also have e.g. something like `netdata --detect` to run it manually).

I would suggest this isn't really relevant to this PR, so maybe best to just leave the startup dependencies for the time being and create a new issue to track it?

Cheers,
Matthew

**Comment by ktsaou** [comment]
> I would suggest this isn't really relevant to this PR, so maybe best to just leave the startup dependencies for the time being and create a new issue to track it?

you are right!

**Comment by mcnewton** [comment]
@ktsaou I have created issue https://github.com/firehol/netdata/issues/170 to track this.

**Comment by mcnewton** [comment]
Updated debian package again.

* integrate security concerns from https://github.com/firehol/netdata/issues/157 - apps.plugin is now installed root:netdata 0754 and has capabilities cap_dac_read_search,cap_sys_ptrace. So _there is no longer a suid binary :-)_
* move docs to /usr/share/doc/netdata as per standard Debian practise/policy.


Lintian warnings now

* no netdata manpage
* font issues
* node.d/README.md is empty

@ktsaou should node.d/README.md be removed if it has no content? Or maybe a very short description of what the node.d stuff is for? I can't fill it in I'm afraid as I don't know.

Tested successfully on Debian jessie and wheezy.

File: contrib/debian/netdata.service [hunk 15]
```diff
@@ -0,0 +1,13 @@
+[Unit]
```
Comment (alonbl) [inline contrib/debian/netdata.service]: there is no need to duplicate the service, please try to consolidate with what we have at /system

File: system/netdata.logrotate.in [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+	missingok
+	rotate 14
+	compress
+	delaycompress
+	notifempty
+	sharedscripts
+	postrotate
+		if service netdata status > /dev/null ; then \
+			service netdata restart > /dev/null; \
+		fi;
```
Comment (alonbl) [inline system/netdata.logrotate.in]: well it will not work with openrc, I guess we need two files if we do not use copytruncate.

but restarting the service will damage user interface, no? do we really want to restart?

File: contrib/debian/netdata.service (hunk 15 above)
Comment (mcnewton) [inline contrib/debian/netdata.service]: RedHat systems don't use /etc/default, but something like /etc/sysconfig. So you'd end up with trying to substitute/remove lines from a single unit file to make it match. The dependency files are all different in .rpm systems as well (e.g. http.service rather than apache2.service), though I removed them from the Debian unit file from the above discussion.

Most other package I've seen ship different unit files for .deb vs .rpm systems for this very reason. Unless you can see a way to cleanly contain the differences. I would doubt it's worth it.

File: system/netdata.logrotate.in (hunk 16 above)
Comment (shartge) [inline system/netdata.logrotate.in]: I think netdata needs a `reload` option (normally triggered via SIGHUP) which only reopens the log files but does not restart any processes. copytruncate is an ugly hack which normally should be avoided unless the software really can't be changes to behave correctly (i.e. most enterprise Java solutions).

Comment (mcnewton) [inline system/netdata.logrotate.in]: restart here gives the tiniest break in the web graphs. Might be slightly annoying I guess, but many things have a very brief break in service at 06:25 when the logs rotate, so it's not unusual.

The best solution would be to signal to the daemon to reopen its log files (-HUP, -USR1, etc). Admit I haven't looked to see if it will currently do this, though.

**Comment by lhw** [comment]
You guys seem to have this under control. Some suggestions I wanted to make though:

- source/format native is only used for packages which have no purpose outside of debian: [Wiki](https://wiki.debian.org/DebianMentorsFaq#What_is_the_difference_between_a_native_Debian_package_and_a_non-native_package.3F)
- I think it would make sense to separate the native binaries and the data into different packages to remove unnecessary duplication on the ftps. e.g. netdata (multiarch: foreign) and netdata-data (all). You could even separate the plugins into different packages so you can have a package with the nodejs plugins without requiring nodejs on the main package.
- I also think it would be easier to maintain the debian directory separately from the upstream repository and base it on the tags/releases instead of HEAD.

File: contrib/debian/rules [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+%:
+	# For jessie and beyond
+	dh $@ --with systemd
+	# For wheezy use the following (and also remove
+	# dh-systemd Build-Depends from the control file)
+	#dh $@
+
+override_dh_auto_configure:
+	# Build the configure script if it doesn't already exist
+	-[ ! -x $(CURDIR)/configure ] && $(CURDIR)/autogen.sh
```
Comment (terryburton) [inline contrib/debian/rules]: Does [dh-autoreconf](https://wiki.debian.org/Autoreconf#Using_dh-autoreconf) take care of this? Untested, so perhaps I've missed the point.

```
%:
	dh $@  --with autoreconf --with=systemd
```

Requires build-depend on dh-autoreconf.

Comment (lhw) [inline contrib/debian/rules]: That should indeed work. And dh-autoreconf depends on the automake packages so those two extra dependencies can be removed.

File: contrib/debian/control [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+Build-Depends: debhelper (>= 9),
+ dpkg-dev (>= 1.13.19),
+ dh-systemd (>= 1.5),
+ zlib1g-dev,
+ autoconf,
+ automake,
+ pkg-config
+Section: net
+Priority: optional
+Maintainer: Costa Tsaousis <costa@tsaousis.gr>
```
Comment (lhw) [inline contrib/debian/control]: This should be the debian maintainer name. Upstream author is in debian/copyright

**Comment by ktsaou** [comment]
When this is ready for merge, please let me know.

**Comment by mcnewton** [comment]
Sorry all, not particularly well and feeling really rough at the moment. I'll take a look at the outstanding things as soon as I'm able.

File: contrib/debian/rules (hunk 17 above)
Comment (mcnewton) [inline contrib/debian/rules]: @terryburton yes works fine - apart from when building with sbuild you have to have dh-autoreconf installed on the main system as well, and if it isn't it just bombs out with an error. But 99.9% of people won't be doing that, so it shouldn't matter :)

@lhw confirmed - extra dependencies can be removed. Thanks.

File: contrib/debian/control (hunk 18 above)
Comment (terryburton) [inline contrib/debian/control]: @lhw Agreed. Perhaps an obvious example name will do.

RE:
> I also think it would be easier to maintain the debian directory separately from the upstream repository and base it on the tags/releases instead of HEAD.

I've found it useful to use the main repository as a collaboration point to promote consistent packaging across distributions and to allow the user to easily generate their own personal packages, as seems to be the case here. Distribution packagers then have the option of using the example packaging as a basis for their own work, or not.

I've done this with [BWIPP](https://github.com/bwipp/postscriptbarcode) making it clear that the packaging is provided as a distribution-agnostic example:

https://github.com/bwipp/postscriptbarcode/tree/master/packaging-examples

I use OBS to check out of the main repository and build the example packaging into binaries for most major distributions on each commit (or tag) using a trigger as part of my Continuous Integration so that I know when I've broken something or made a packager's life difficult - in which case I fix things and provide heads up.

https://build.opensuse.org/package/show/home:terryburton:postscriptbarcode/libpostscriptbarcode

(A nit is that OBS makes it difficult to not have the debian packaging in the repository root as debian/ otherwise it would go in example-packaging/debian/ or at lease debian.upstream/. I'm working on a fix for this.)

I absolutely agree that the actual distribution-specific packaging should be hosted wherever is least-surprising to the community, e.g. Alioth, etc.

Comment (mcnewton) [inline contrib/debian/control]: @lhw thanks for the comments!

You're right on all of it, but I think we're looking at the packaging from two different perspectives.

If you're building a debian/ directory for the Debian project, then yes add the other dependencies etc and split up into multiple packages. You can deprecate older systems because you won't generally need to build for them.

But this is putting together packaging so that someone can take the upstream source, hopefully on pretty much any recent .deb based system, and build something that works. That's a different goal (and why we moved it into contrib/).

So although I haven't tested it, hopefully you should be able to download netdata source onto Ubuntu whatever, or Mint, or other systems and easily build a recent package for your system. Adding in all the Debian dependencies or strictly following Debian Policy on this doesn't help them, as their systems are likely to be different. This packaging should be as simple and universal as possible.

That's also why source/format is "native". Non-native means someone who checks out from git then has to build a tarball (following a precise naming scheme) in the parent directory before the build will even commence. Whereas "native" is perfect for a non-official package - the build will complete for them and they don't need to care about generating a diff (which is empty anyway).

Does that make sense?

However - very happy if you want to take this as a basis for an official Debian package, host it on the Debian systems as you normally would and then of course it would be put in the standard debian/ location as you build.

Having said that, "wrap-and-sort" - cool :-) not come across that before!

Cheers,
Matthew

File: contrib/debian/rules (hunk 17 above)
Comment (mcnewton) [inline contrib/debian/rules]: @terryburton though a small note that it's not completely clean on wheezy - you still need to have a build-depends on pkg-config. I've just added a README to explain for anyone wanting to do that.

File: contrib/debian/control (hunk 18 above)
Comment (lhw) [inline contrib/debian/control]: @mcnewton Makes perfect sense that you want to target as many distributions as possible. That explains what I see here.

I think I will take ownership of the RFP bug on debian and start working on getting it into unstable. Your contributions will make a great base to start on.

File: contrib/debian/netdata.service [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+Description=netdata real-time system monitoring
+After=network.target
+
+[Service]
+Type=simple
+EnvironmentFile=-/etc/default/netdata
+ExecStart=/usr/sbin/netdata -nd $EXTRA_OPTS
+TimeoutStopSec=30
+Restart=5s
```
Comment (shartge) [inline contrib/debian/netdata.service]: I must admit, I made a mistake here. `Restart=5s`is totally wrong, the correct lines look like this:
```
Restart=always
RestartSec=5
```

Comment (mcnewton) [inline contrib/debian/netdata.service]: @shartge thanks - I didn't think it looked quite right, but then forgot to check. It works fine with this now. I'll push a new version with this in it soon.

**Comment by mcnewton** [comment]
Pushed latest updates:

* fixed up systemd Restart, thanks @shartge
* "static" web files now in /usr/share/netdata
* html files now in /var/lib/netdata/www
* correctly sets web file group on upgrade as well as install
* web files now owned root:netdata (8bb28f2f6a6526847fc3d2bc8a216e041a0f8677 - thanks @ktsaou)
* files in /var/lib/netdata/www now set as conffiles so an upgrade doesn't clobber them if they have been changed

Tested OK in jessie and wheezy sbuild environments.

Hoping we might finally be getting close! :)

File: system/netdata.logrotate.in (hunk 16 above)
Comment (ktsaou) [inline system/netdata.logrotate.in]: > The best solution would be to signal to the daemon to reopen its log files (-HUP, -USR1, etc). Admit I haven't looked to see if it will currently do this, though.

Hm... I don't know if this can be done. The whole process tree of netdata has `stderr` attached to `error.log`. Most children (bash, node.js, anything) are not aware of this. So, even if the daemon re-opens its files, the children will not.

Comment (mcnewton) [inline system/netdata.logrotate.in]: Do the children write to the logfiles in /var/log/netdata as well, though? It's just access.log, debug.log and error.log.

Comment (ktsaou) [inline system/netdata.logrotate.in]: Indirectly yes: when the daemon spawns a child, it leaves `/var/log/netdata/error.log` attached to the process `stderr`. So, let's say we have a bash plugin. This command:

```sh
echo >&2 "something failed"
```

will send this message to `/var/log/netdata/error.log`.
But the child is not aware of this. And the daemon is not in the middle anymore (as far as error.log is concerned).

So, we can do somehting in the daemon (SIGUSR1 maybe), but still `error.log` will be a problem.

Comment (mcnewton) [inline system/netdata.logrotate.in]: Ah, OK, got you.

I don't think restart is that bad (and definitely not to get worried about now). e.g. on my system it leaves a gap in the graphs of just 2 seconds.

**Comment by TimRots** [comment]
any updates on when/if this PR will be merged?

**Comment by ktsaou** [comment]
I will press the button once @mcnewton  says so...

**Comment by mcnewton** [comment]
Sorry @ktsaou - forgot! Yes, this should be good to go.

At this stage any problems are likely to be where people use systems that I've not been able to test on, such as Ubuntu. But they can easily be fixed after it's merged if needed.

Thanks,
Matthew

**Comment by ktsaou** [comment]
merged!

**Comment by TimRots** [comment]
Thank you both for the great efforts.

**Comment by lhw** [comment]
@ktsaou Is there a plan to tag a new release version in the near future?

**Comment by ktsaou** [comment]
I'll try to do it this evening.

**Comment by lhw** [comment]
Great! I'll start getting this into debian asap then. Shouldn't be hard with the great basis already provided by @mcnewton.

**Comment by ktsaou** [comment]
Released v1.1.0 !
Thanks!

**Comment by joelsdc** [comment]
Just adding some extra info here....

On v1.1.0, when you build the debian package, it works but the file name and all version references say v1.0.0.

Thanks for the effort! All your work is really appreciated.

**Comment by lhw** [comment]
@joelsdc the debian package version is solely based on the version mentioned in the changelog: https://github.com/firehol/netdata/blob/master/contrib/debian/changelog#L1
Unless the version there is updated it will show the latest version in the file it can find.

**Comment by ktsaou** [comment]
oops! I didn't know that and I forgot to update the changelog at the latest release.
Can I fix it now or a new release is needed (which will happen either way in the next few days)?

**Comment by lhw** [comment]
Don't need to update it right now. The changelog for the actual debian release is maintained in our package git anyway. We just used the contrib/debian as base.
